src/osd/scheduler/mClockScheduler.h

```diff
@@ -113,15 +119,90 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
       crimson::dmclock::ClientInfo(1, 1, 1)
     };
 
+    // Mutex to protect ClientRegistry operations
+    mutable std::shared_mutex rw_mtx;
+    using WriteGuard = std::unique_lock<decltype(rw_mtx)>;
+    using ReadGuard = std::shared_lock<decltype(rw_mtx)>;
+
+    // Client info Maps and accessors
     crimson::dmclock::ClientInfo default_external_client_info = {1, 1, 1};
     std::map<client_profile_id_t,
 	     crimson::dmclock::ClientInfo> external_client_infos;
+    std::map<client_profile_id_t,
+             counter_t> external_client_infos_tracker;
     const crimson::dmclock::ClientInfo *get_external_client(
       const client_profile_id_t &client) const;
+    void set_client_infos_tracker(const client_profile_id_t &id);
+
+    // External client registry clean-up job
+    std::unique_ptr<crimson::RunEvery> cleaning_job;
+    using TimePoint = decltype(std::chrono::steady_clock::now());
+    using MarkPoint = std::pair<TimePoint, counter_t>;
+    std::deque<MarkPoint> clear_mark_points;
+    // Frequency of the clean-up job
+    Duration clear_every = default_clear_every;
+    // Max age of a client entry in the ClientRegistry
+    Duration clear_age = default_clear_age;
+    // Max number of clients to clear at a time
+    unsigned clear_max = default_clear_max;
+    // Every add/update to the ClientRegistry increments a tick for a client
+    counter_t tick = 0;
   public:
     void update_from_config(const ConfigProxy &conf);
+
     const crimson::dmclock::ClientInfo *get_info(
       const scheduler_id_t &id) const;
+
+    void set_info(const scheduler_id_t &id,
+      const client_qos_params_t &qos_profile_params);
+
+    void clear_info(const scheduler_id_t &id);
+
+    unsigned get_external_client_infos_size() {
+      ReadGuard g(rw_mtx);
+      return external_client_infos.size();
+    }
+
+    unsigned get_external_client_infos_tracker_size() {
+      ReadGuard g(rw_mtx);
+      return external_client_infos_tracker.size();
+    }
+
+    void set_clear_every(Duration _clear_every) {
+      WriteGuard g(rw_mtx);
+      if (cleaning_job) {
+        cleaning_job->try_update(
+          std::chrono::duration_cast<std::chrono::milliseconds>(_clear_every));
+        clear_every = _clear_every;
+      }
+    }
+
+    Duration get_clear_every() {
+      ReadGuard g(rw_mtx);
+      return clear_every;
+    }
+
+    void set_clear_age(Duration _clear_age) {
+      WriteGuard g(rw_mtx);
+      clear_age = _clear_age;
+    }
+
+    Duration get_clear_age() {
+      ReadGuard g(rw_mtx);
+      return clear_age;
+    }
+
+    void do_clean();
+
+    void init_cleaning_job() {
+      if (!cleaning_job) {
+        cleaning_job = std::unique_ptr<crimson::RunEvery>(
+          new crimson::RunEvery(get_clear_every(),
+                                std::bind(&ClientRegistry::do_clean,
+                                this)));
```

Comment:
RunEvery looks like another implementation of SafeTimer.  Rather than auditing RunEvery, I'd really rather we just stick to SafeTimer.

Reply:
Makes sense. I have re-implemented the same functionality with SafeTimer. I had to accordingly replace the std::shared_mutex with ceph::mutex.